src/pcp.rs

```diff
@@ -523,8 +533,17 @@ where
     G: Gadget<F>,
     V: Value<F, G>,
 {
-    let e = x.gadget(0).call(&vf.f_at_r)?;
-    if e == vf.p_at_r && vf.v == F::zero() {
+    let mut g = x.gadget(0);
+    let vf_len = vf.data.len();
+    if vf_len < 2 + g.call_in_len() {
+        return Err(PcpError::Decide("short verifier message"));
+    }
+
+    let v = vf.data[0];
+    let f_at_r = &vf.data[1..vf_len - 1];
+    let p_at_r = vf.data[vf_len - 1];
```

Comment:
I'd prefer to have methods `Verifier::validity(&self) -> F`, `Verifier::p_at_r(&self) -> F` and `Verifier::f_at_r(&self) -> &[F]` so that the implementation detail of how `Verifier` stores its data is abstracted away. Similarly, `Verifier::new(v: F, p_at_r: F, f_at_r: &[F]) -> Verifier<F>` would be better than making `query()` construct the `Vec<F>`.

Reply:
- Added `validity()`, `f_at_r()`, and `p_at_r()` to Verifier implementation.
- I'm not sure what you mean for the second part. `query()` returns a `Verifier<F>`, not a `Vec<F>`.

Reply:
>     * I'm not sure what you mean for the second part. `query()` returns a `Verifier<F>`, not a `Vec<F>`.

What I mean is that instead of 
```
    Ok(Verifier {
            data: verifier_data,
   })
```
`query()` could do
```
Ok(Verifier::new(v, p_at_r, f_at_r))
```
And `Verifier::new()` would be responsible for allocating an appropriately sized `Vec<F>` and constructing a `Verifier` from it. This would also absolve `decide()` of the need to check that the `data` member of the `Verifier` it is passed has an appropriate length.

Reply:
I see. My one argument against is that it's slightly less efficient because it requires re-allocating an length-`(2+l)` vector and copying data into it. Usually `l` will be small, but not always.

In any case, IMO we're talking about an implementation detail that (hopefully) isn't exposed to the user. From the user's perspective, a `Verifier<F>` is constructed from a `Vec<F>` of any length. (Of course, not all lengths are valid, which `decide()` needs to check for.) I'm not sure I see the value in exposing more detail than that in the API.

Reply:
True, it is all internal to the `pcp` module.